lisa/_assets/kmodules/lisa/fs.c

```diff
@@ -0,0 +1,755 @@
+
+		if (size >= MAX_BUF_SIZE)
+			size = MAX_BUF_SIZE - 1;
+
+		if (copy_from_user(kbuf, buf + done, size)) {
+			kfree(kbuf);
+			goto done;
+		}
+		kbuf[size] = '\0';
+		s = sep = kbuf;
+		do  {
+			sep = strchr(s, ',');
+			if (sep) {
+				*sep = '\0';
+				++sep;
+				++done;
+			}
+			done += size = strlen(s);
+			/* skip leading whitespaces */
+			while (isspace(*s) && *(s++))
+				--size;
+			if (!*s)
+				goto next;
+			if (done < count && !sep) {
+				/* carry over ... */
+				done -= strlen(s);
+				goto next;
+			}
+			/* skip trailing whitespaces */
+			while (size && isspace(s[--size]));
+			if (strlen(s) > ++size)
+				s[size] = '\0';
+
+			ret = feature_param_add_new(entry, s);
+			if (ret) {
+				done = ret;
+				goto done;
+			}
+
+			if (ppos)
+				*ppos += 1;
+		next:
+			s = sep;
+		} while (s);
+	}
+done:
+	kfree(kbuf);
+leave:
+	mutex_unlock(&interface_lock);
+	return done;
+}
+
+int lisa_param_feature_release(struct inode *inode, struct file *file)
+{
+	return file->f_mode & FMODE_READ ? seq_release_private(inode, file) : 0;
+}
+
+static struct file_operations lisa_param_feature_fops = {
+	.owner = THIS_MODULE,
+	.open = lisa_param_feature_open,
+	.read = seq_read,
+	.write = lisa_param_feature_write,
+	.release = lisa_param_feature_release,
+};
+
+/////////////////////////////////////
+//           configs
+/////////////////////////////////////
+
+/* TODO:
+ * linux:
+ * int (*mkdir) (struct mnt_idmap *, struct inode *,struct dentry *,
+ *		      umode_t);
+ * android:
+ * int (*mkdir) (struct user_namespace *, struct inode *,struct dentry *,
+ *                    umode_t);
+ */
+
+static int lisa_fs_syscall_mkdir(struct mnt_idmap *idmap,
+				 struct inode *inode, struct dentry *dentry,
+				 umode_t mode)
+{
+	struct dentry *my_dentry;
+	struct lisa_cfg *cfg;
+	int ret;
+
+	cfg = allocate_lisa_cfg(dentry->d_name.name);
+	if (!cfg)
+		return -ENOMEM;
+
+	mutex_lock(&interface_lock);
+
+	my_dentry = lisa_fs_create_single(dentry->d_parent, dentry->d_name.name,
+					  &simple_dir_inode_operations,
+					  &simple_dir_operations,
+					  S_IFDIR | mode, cfg);
+	if (!my_dentry)
+		goto error;
+
+	ret = init_lisa_cfg(cfg, &cfg_list, my_dentry);
+	if (ret)
+		return -ENOMEM;
+
+	lisa_fs_create_files(my_dentry, false, cfg);
+	mutex_unlock(&interface_lock);
+	return 0;
+
+error:
+	free_lisa_cfg(cfg);
+	mutex_unlock(&interface_lock);
+	return ret;
+}
+
+void lisa_fs_remove(struct dentry *dentry)
+{
+	simple_recursive_removal(dentry, NULL);
+}
+
+static int lisa_fs_syscall_rmdir(struct inode *inode, struct dentry *dentry)
+{
+	struct lisa_cfg *cfg;
+	cfg = inode->i_private;
+
+	inode_unlock(inode);
+	inode_unlock(d_inode(dentry));
+
+	mutex_lock(&interface_lock);
+
+	cfg = find_lisa_cfg(dentry->d_name.name);
+	if (!cfg)
+		pr_err("Failed to find config: %s\n", dentry->d_name.name);
+
+	free_lisa_cfg(cfg);
+
+	mutex_unlock(&interface_lock);
+
+	inode_lock_nested(inode, I_MUTEX_PARENT);
+	inode_lock(d_inode(dentry));
+
+	return 0;
```

Comment:
Is this actually removing the directory ?
Looking at simple_rmdir it seems that it might be missing smth , like drop_nlink calls ?

Reply:
The dir removal is done via:
```
lisa_fs_rmdir()
\-free_lisa_cfg()
  \-lisa_fs_remove()
    \-simple_recursive_removal()
```
I.e. the `lisa_cfg` is freed, and while freeing it, its dentry is freed.
This allows to remove the directory associated to a config whenever we free a `lisa_cfg` structure.